### DISCUSSION TITLE
Windows Support

### DISCUSSION THREAD

**Discussion by nik-rev** [TheOdinProject/curriculum#28718] [Ideas]
I am fully aware of the fact that windows support has been requested many times and rejected in the past.

But, out of curiosity, I tried completing all **Installation** instructions in the Foundations course using Windows Powershell, no WSL.

Not only was it an *good* experience, but it can also be quite close to what you'd experience on Linux. Windows users can use package managers like `scoop` instead of `apt-get` in Ubuntu.

The GNU Coreutils are available as a package (1 command away) on Windows which allows us to have commands like `mv`, `touch`, `rm` that are the same as on linux (amongs hundreds of others), which means adding Windows support will be easier

Adding support for Windows won't be nearly as much of a challenge as *I imagined* you imagined it to be, if that makes sense. It could even *reduce* the amount of maintenance load because WSL and VM won't have to be supported anymore.

I may very well be wrong, but i think this is at least worthy of consideration. It's impossible to know for sure, but I wouldn't be surprised if the lack of *native* windows support stops **a lot** of learners from progressing in this course. I want this to be accessible to as many people, and adding native support in Windows would be amazing. 

With that being said, here are the steps to completing Foundations installation (**this is for demo purposes only for the maintainers**)

- All commands are executed in powershell

**Installing scoop package manager**

```powershell
Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Scope CurrentUser
Invoke-RestMethod -Uri https://get.scoop.sh | Invoke-Expression
```

Installing linux commands on windows like `rm`, `mv` and others (this isn't necessary but it's to make the addition of windows support easier in theory for the maintainers, to maintain)

```powershell
scoop bucket add main
scoop install main/uutils-coreutils
```

Some of the installation instructions ask learners to use some commands like `mv` and `rm`, so this package would mean those instructions won't have to be translated to Windows specific commands

### Installations

Installing Google Chrome
```powershell
scoop bucket add extras
scoop install extras/googlechrome
```

Installing VSCode
```powershell
scoop install vscode
```

### Command Line Basics lesson

This lesson will have to have a separate assignment section for windows users. However, pretty much of the commands taught in the assignment are also available on windows in the `coreutils` package. 

This is the only lesson that will have to have some non-minor changes applied to it.

### Setting Up Git

```powershell
scoop install git
```

Copy pasting the following command will automatically generate the SSH key that the learner can copy-paste into their github (we can obviously split the command up into several commands for learning purposes)

- Creates effectively `~/.ssh` directory (using tilde to represent `%userprofile%`
- Generates the ssh key public and private pair
- Copied the public key into the clipboard
- Now the learner can login into github and paste their ssh keys
- All other steps with github remain the same

```powershell
New-Item -ItemType Directory -Path $env:USERPROFILE\.ssh -Force; ssh-keygen -t ed25519 -f "$env:USERPROFILE\.ssh\id_ed25519" -N '""' ; type "$env:USERPROFILE\.ssh\id_ed25519.pub" | clip
```

`git` commands will work since we have installed `git`

### Git Lessons, Git Config, Pushing first project to github...

Content here does not need to change, the `git` commands are going to be the same

### Node.js installation

Includes node, npm and nvm

```powershell
scoop install nodejs
```

### TDD Exercises

Cloning repository, installing jest, etc all can be almost exactly done the same

### Etc

Odin Project has **significantly** more non-"Installations" material than any other material. I think adding support for Windows would not be extremely difficult. 

It could make things *easier* because it would eliminate the need for WSL and Virtual Machine support which from what I've seen are common pain points.

Making the course accessible to just so many more people could potentially be worth making the change. but of course, I'm just making a proposition. 😃

**Comment by bycdiaz** [comment]
There is more for us to consider than the technical ability to install things on windows. If that were the only challenge, Windows would have been supported long ago. But I’m eager to hear from more of our maintainers.

**Comment by QuizeXxx** [comment]
does anyone know, how can i download smth like list of my past ctrl+c?? Cause im really newbie on this site

**Reply by mao-sz** [reply]
This is not the right place to ask, as this is a repo discussion for a specific topic. I'd advice googling for how to do what you want.

**Comment by mao-sz** [comment]
Agreed with Carlos that there is a lot more that goes into us not supporting Windows than "can't install some things" (even if it was just that, there's a lot more to the curriculum than just Foundations - we need to be able support it for *everything*). We also need to be mindful of how many changes would need to be made throughout the entire curriculum and bloating lessons with a bunch of potentially completely different instructions just because of another OS, and one where things work very differently, and more often than not, you'll be working in a Unix environment of some sort in the professional sphere. If someone does end up working with Windows, they'll be equipped to learn what they need to learn - we don't *have* to include support for it.

So I'm on the side of keeping our OS support as is, but thank you for making the suggestion and doing some digging.
